pages/api/discover/popular.js

```diff
@@ -16,25 +19,29 @@ export default async function handler(req, res) {
 
   const directoryPath = path.join(process.cwd(), "data");
 
-  const fullPopularProfiles = popularProfiles.flatMap((profile) => {
-    const filePath = path.join(directoryPath, `${profile.username}.json`);
-    try {
-      const user = JSON.parse(fs.readFileSync(filePath, "utf8"));
-
-      if (user.displayStatsPublic) {
-        return {
-          ...user,
-          ...profile._doc,
-        };
+  const fullPopularProfiles = await Promise.all(
+    popularProfiles.flatMap(async (profile) => {
+      const filePath = path.join(directoryPath, `${profile.username}.json`);
+      try {
+        const user = await memoizedReadAndParseJsonFile(filePath);
```

Comment:
I think the await will be blocking and the `Promise.all` needs an array of promises to resolve to be beneficial

Reply:
Wouldn't the benefit be that the `memoizedReadAndParseJsonFile` would be executed concurrently instead of sequentially?

Reply:
Not if there is an await inside the promise all, the await should be outside because the promise all wants to accept everything async and resolve it concurrently

Reply:
From what I see, `popularProfiles.flatMap(async fn)` returns a list of promises, wrapping it up in `Promises.all` then blocks waiting for all the promises in that list to resolve.

```
linkfree  | [
linkfree  |   Promise { <pending> }, Promise { <pending> },
linkfree  |   Promise { <pending> }, Promise { <pending> },
linkfree  |   Promise { <pending> }, Promise { <pending> },
linkfree  |   Promise { <pending> }, Promise { <pending> },
linkfree  |   Promise { <pending> }, Promise { <pending> },
linkfree  |   Promise { <pending> }, Promise { <pending> },
linkfree  |   Promise { <pending> }, Promise { <pending> },
linkfree  |   Promise { <pending> }, Promise { <pending> },
linkfree  |   Promise { <pending> }, Promise { <pending> }
linkfree  | ]
```
Above is the output of `popularProfiles.flatMap(async fn)`, before resolving using Promises.all

Reply:
> > The Promise.all() static method takes an iterable of promises as input and returns a single Promise
> 
> I think the await will be blocking and the `Promise.all` needs an array of promises to resolve to be beneficial

From what I see, `popularProfiles.flatMap(async fn)` returns a list of promises, wrapping it up in `Promises.all` then blocks waiting for all the promises in that list to resolve.

```
linkfree  | [
linkfree  |   Promise { <pending> }, Promise { <pending> },
linkfree  |   Promise { <pending> }, Promise { <pending> },
linkfree  |   Promise { <pending> }, Promise { <pending> },
linkfree  |   Promise { <pending> }, Promise { <pending> },
linkfree  |   Promise { <pending> }, Promise { <pending> },
linkfree  |   Promise { <pending> }, Promise { <pending> },
linkfree  |   Promise { <pending> }, Promise { <pending> },
linkfree  |   Promise { <pending> }, Promise { <pending> },
linkfree  |   Promise { <pending> }, Promise { <pending> }
linkfree  | ]
```
Above is the output of `popularProfiles.flatMap(async fn)`, before resolving using Promises.all

Reply:
From the code it returns an object on line `29`

```js
return {
             ...user,
             ...profile._doc,
           };
```